cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -80,6 +80,134 @@ const FunctionDoc list_value_length_doc{
+  using IndexScalarType = typename TypeTraits<IndexType>::ScalarType;
+  const auto& index_scalar = batch[1].scalar_as<IndexScalarType>();
+  if (ARROW_PREDICT_FALSE(!index_scalar.is_valid)) {
+    return Status::Invalid("Index must not be null");
+  }
+  ListArrayType list_array(batch[0].array());
+  auto index = index_scalar.value;
+  if (ARROW_PREDICT_FALSE(index < 0)) {
+    return Status::Invalid("Index ", index,
+                           " is out of bounds: should be greater than or equal to 0");
+  }
+  std::unique_ptr<ArrayBuilder> builder;
+  RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), list_array.value_type(), &builder));
+  RETURN_NOT_OK(builder->Reserve(list_array.length()));
+  for (int i = 0; i < list_array.length(); ++i) {
+    if (list_array.IsNull(i)) {
+      RETURN_NOT_OK(builder->AppendNull());
+      continue;
+    }
+    std::shared_ptr<arrow::Array> value_array = list_array.value_slice(i);
+    auto len = value_array->length();
+    if (ARROW_PREDICT_FALSE(index >= static_cast<typename IndexType::c_type>(len))) {
+      return Status::Invalid("Index ", index, " is out of bounds: should be in [0, ", len,
+                             ")");
+    }
+    RETURN_NOT_OK(builder->AppendArraySlice(*value_array->data(), index, 1));
+  }
+  ARROW_ASSIGN_OR_RAISE(auto result, builder->Finish());
+  out->value = result->data();
+  return Status::OK();
+}
+
+template <typename, typename IndexType>
+struct ListElementScalar {
+  static Status Exec(KernelContext* /*ctx*/, const ExecBatch& batch, Datum* out) {
+    using IndexScalarType = typename TypeTraits<IndexType>::ScalarType;
+    const auto& index_scalar = batch[1].scalar_as<IndexScalarType>();
+    if (ARROW_PREDICT_FALSE(!index_scalar.is_valid)) {
+      return Status::Invalid("Index must not be null");
+    }
+    const auto& list_scalar = batch[0].scalar_as<BaseListScalar>();
+    if (ARROW_PREDICT_FALSE(!list_scalar.is_valid)) {
+      out->value = MakeNullScalar(
+          checked_cast<const BaseListType&>(*batch[0].type()).value_type());
+      return Status::OK();
+    }
+    auto list = list_scalar.value;
+    auto index = index_scalar.value;
+    auto len = list->length();
+    if (ARROW_PREDICT_FALSE(index < 0 ||
+                            index >= static_cast<typename IndexType::c_type>(len))) {
+      return Status::Invalid("Index ", index, " is out of bounds: should be in [0, ", len,
+                             ")");
+    }
+    ARROW_ASSIGN_OR_RAISE(out->value, list->GetScalar(index));
+    return Status::OK();
+  }
+};
+
+template <typename InListType, typename IndexType>
+void AddListElementArrayKernel(ScalarFunction* func) {
+  auto inputs = {InputType::Array(InListType::type_id),
+                 InputType::Scalar(IndexType::type_id)};
+  auto output = OutputType{ListValuesType};
+  auto sig =
+      KernelSignature::Make(std::move(inputs), std::move(output), /*is_varargs=*/false);
+  auto array_exec = ListElementArray<InListType, IndexType>;
```

Comment:
Note if you wanted, for consistency, this could be `GenerateInteger<ListElementArray, InListType>`.